### PR TITLE
Add API to probe the logstream batch writer if more bytes can be written without writing them

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR adds a new API method, `LogStreamBatchWriter#canWriteAdditionalEvent(int)`. This allows users of the writer to probe if adding the given amount of bytes to the batch would cause it to become un-writable, without actually having to write anything to the batch, or even modify their DTO (e.g. the `TypedRecord<?>` in the engine).

To avoid having dispatcher details leak into the implementation, an analogous method is added to the dispatcher, `Dispatcher#canClaimFragmentBatch(int, int)`, which will compare the given size, framed and aligned, with the max fragment length. This is the main building block to eventually solve #5525, and enable other use cases (e.g. multi-instance creation) which deal with large batches until we have a more permanent solution (e.g. chunking follow up batches).

NOTE: the tests added in the dispatcher are not very good, but I couldn't come up with something else that wouldn't be too coupled to the implementation (i.e. essentially reusing `LogBufferAppender`). I would like some ideas/suggestions.

NOTE: this PR comes out of the larger one, #8491. You can check that one out to see how the new API would be used, e.g. in the `JobBatchCollector`. As such, this is marked for backporting, since we'll backport the complete fix for #5525.

## Related issues

related to #5525 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

### CODE REVIEW

File: dispatcher/src/test/java/io/camunda/zeebe/dispatcher/DispatcherTest.java [hunk 1]
```diff
@@ -133,6 +133,34 @@ public void shouldNotClaimBeyondPublisherLimit() {
...
+    final int batchLength = dispatcher.getMaxFragmentLength();
+
+    // when
+    final var canClaimFragmentBatch = dispatcher.canClaimFragmentBatch(fragmentCount, batchLength);
+
+    // then
+    assertThat(canClaimFragmentBatch)
+        .as("cannot claim when the unframed, unaligned batch is the max fragment length")
+        .isFalse();
+  }
```
Comment (saig0) [inline dispatcher/src/test/java/io/camunda/zeebe/dispatcher/DispatcherTest.java:162]: :wrench: We could write a more general test case to verify that we can claim a batch.

The current test case covers a corner case. As a developer, if I don't know the implementation then the test case doesn't make much sense (i.e. why can I claim `2 * max/2` but not `1 * max`?). 

We could leave this test case to cover the corner case. But it may be helpful to write a more obvious test case (e.g. claim `2 * max` or `3 * max/2`).

**Review by saig0 — APPROVED** [review]
@npepinpe looks good so far :+1: 
I'm not so happy about the writer test but I have no better idea :sweat_smile: 

Please have a look at my comment. But feel free to merge :rocket:

File: dispatcher/src/test/java/io/camunda/zeebe/dispatcher/DispatcherTest.java (hunk 1 above)
Comment (npepinpe) [reply dispatcher/src/test/java/io/camunda/zeebe/dispatcher/DispatcherTest.java]: I'm still not happy with the tests, but I agree with your point, it makes sense to distinguish both cases :slightly_smiling_face:

**Comment by npepinpe** [comment]
I also took the opportunity to update the `DispatcherTest` to Junit 5, as it was an extremely easy thing to do and we do want to migrate to Junit 5 completely eventually. I don't think it needs another review, but I'll ping you on slack for a quick chance to veto anyway :wink:

**Comment by npepinpe** [comment]
bors merge
